instana/resource-sli-config.go

```diff
@@ -0,0 +1,261 @@
+func NewSliConfigResourceHandle() *ResourceHandle {
+	return &ResourceHandle{
+		ResourceName: ResourceInstanaSliConfig,
+		Schema: map[string]*schema.Schema{
+			SliConfigFieldName:                       SliConfigName,
+			SliConfigFieldFullName:                   SliConfigFullName,
+			SliConfigFieldInitialEvaluationTimestamp: SliConfigInitialEvaluationTimestamp,
+			SliConfigFieldMetricConfiguration:        SliConfigMetricConfiguration,
+			SliConfigFieldSliEntity:                  SliConfigSliEntity,
+		},
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Type:    sliConfigSchemaV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: sliConfigStateUpgradeV0,
+				Version: 0,
+			},
+		},
+		RestResourceFactory:  func(api restapi.InstanaAPI) restapi.RestResource { return api.SliConfigs() },
+		UpdateState:          updateStateForSliConfig,
+		MapStateToDataObject: mapStateToDataObjectForSliConfig,
+	}
+}
+
+func updateStateForSliConfig(d *schema.ResourceData, obj restapi.InstanaDataObject) error {
+	sliConfig := obj.(restapi.SliConfig)
+
+	metricConfiguration := map[string]string{
+		SliConfigFieldMetricName:        sliConfig.MetricConfiguration.Name,
+		SliConfigFieldMetricAggregation: sliConfig.MetricConfiguration.Aggregation,
+		SliConfigFieldMetricThreshold:   fmt.Sprintf("%f", sliConfig.MetricConfiguration.Threshold),
+	}
+
+	sliEntity := map[string]string{
+		SliConfigFieldSliType:       sliConfig.SliEntity.Type,
+		SliConfigFieldApplicationID: sliConfig.SliEntity.ApplicationID,
+		SliConfigFieldServiceID:     sliConfig.SliEntity.ServiceID,
+		SliConfigFieldEndpointID:    sliConfig.SliEntity.EndpointID,
+		SliConfigFieldBoundaryScope: sliConfig.SliEntity.BoundaryScope,
+	}
+
+	d.Set(SliConfigFieldFullName, sliConfig.Name)
+	d.Set(SliConfigFieldInitialEvaluationTimestamp, sliConfig.InitialEvaluationTimestamp)
+	d.Set(SliConfigFieldMetricConfiguration, metricConfiguration)
+	d.Set(SliConfigFieldSliEntity, sliEntity)
+
+	d.SetId(sliConfig.ID)
+	return nil
+}
+
+func mapStateToDataObjectForSliConfig(d *schema.ResourceData, formatter utils.ResourceNameFormatter) (restapi.InstanaDataObject, error) {
+	metricConfigurationsList := d.Get(SliConfigFieldMetricConfiguration).([]interface{})
+	var metricConfiguration restapi.MetricConfiguration
+	if len(metricConfigurationsList) > 0 {
+		metricConfigurationState := metricConfigurationsList[0].(map[string]interface{})
+		if len(metricConfigurationState) != 0 {
+			metricConfiguration = restapi.MetricConfiguration{
+				Name:        metricConfigurationState[SliConfigFieldMetricName].(string),
+				Aggregation: metricConfigurationState[SliConfigFieldMetricAggregation].(string),
+				Threshold:   metricConfigurationState[SliConfigFieldMetricThreshold].(float64),
+			}
+		}
+	}
+
+	sliEntitiesList := d.Get(SliConfigFieldSliEntity).([]interface{})
+	var sliEntity restapi.SliEntity
+	if len(sliEntitiesList) > 0 {
```

Comment:
could be moved into a separate method mapSliEntityListFromState()

Reply:
refactored into separate method